### PR TITLE
crud steps

### PR DESCRIPTION
**Pull request by pldespaigne**
State issue number in your PR to close it when PR merged: issue #XXX
e.g. : "Resolve the bundle size ❤️: issue #1234"

### CODE REVIEW

File: libs/waterfall/src/lib/components/graph/conditions/conditions.component.ts [hunk 1]
```diff
@@ -16,6 +16,11 @@ export class WaterfallConditionsComponent implements OnInit {
 
   @Input() rightForm: RightForm;
 
+  @Output() createStep = new EventEmitter();
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/conditions/conditions.component.ts:19]: `createStep` is not used

File: libs/waterfall/src/lib/components/graph/conditions/step-item/step-item.component.html [hunk 2]
```diff
@@ -0,0 +1,13 @@
+
+<p>Step {{ index + 1}}</p>
+<p>({{conditionCount}} condition{{ conditionCount > 1 ? 's' : '' }})</p>
+
+<span class="spacer"></span>
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/conditions/step-item/step-item.component.html:5]: you don't need class as this is the only span

File: libs/waterfall/src/lib/components/graph/graph.component.ts [hunk 3]
```diff
@@ -106,7 +113,15 @@ export class WaterfallGraphComponent implements OnInit, OnDestroy {
     right.name = this.rightForm.controls.name.value;
     right.percent = this.rightForm.controls.percent.value;
     right.rightholderId = this.rightholders.find(r => r.name === this.rightForm.controls.org.value)?.id ?? '';
-    right.conditions = { operator: 'AND', conditions: this.rightForm.controls.conditions.value };
+
+    if (right.type === 'vertical') {
+      const steps = changes.updated.rights.filter(r => r.groupId === right.id);
+      this.rightForm.controls.steps.value.forEach((condtions, index) => {
+        steps[index].conditions = { operator: 'AND', conditions: condtions };
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/graph.component.ts:120]: maybe use `createCondition()`
```suggestion
        steps[index].conditions = { operator: 'AND', conditions: condtions.map(c => createCondition(c)) };
```

File: libs/waterfall/src/lib/components/graph/graph.component.ts [hunk 4]
```diff
@@ -106,7 +113,15 @@ export class WaterfallGraphComponent implements OnInit, OnDestroy {
...
+      });
+    } else {
+      right.conditions = { operator: 'AND', conditions: this.rightForm.controls.steps.value[0] };
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/graph.component.ts:123]: same
```suggestion
      right.conditions = { operator: 'AND', conditions: createCondition(this.rightForm.controls.steps.value[0]) };
```

File: libs/waterfall/src/lib/components/graph/graph.component.ts [hunk 5]
```diff
@@ -173,6 +188,46 @@ export class WaterfallGraphComponent implements OnInit, OnDestroy {
...
 
+  async createNewStep() {
+    const graph = this.nodes$.getValue();
+    const rightId = this.selected$.getValue();
+    createStep(rightId, graph);
+    console.log(graph);
+    const newGraph = fromGraph(graph);
+    console.log(newGraph);
+    const changes = computeDiff({ rights: this.rights, sources: this.sources }, newGraph);
+    console.log(changes);
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/graph.component.ts:199]: console log

File: libs/waterfall/src/lib/components/graph/graph.component.ts [hunk 6]
```diff
@@ -173,6 +188,46 @@ export class WaterfallGraphComponent implements OnInit, OnDestroy {
...
+
+  async deleteStep(index: number) {
+    const graph = this.nodes$.getValue();
+    const rightId = this.selected$.getValue();
+    deleteStep(rightId, index, graph);
+    console.log(graph);
+    const newGraph = fromGraph(graph);
+    console.log(newGraph);
+    const changes = computeDiff({ rights: this.rights, sources: this.sources }, newGraph);
+    console.log(changes);
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/graph.component.ts:219]: console.log

File: libs/waterfall/src/lib/components/graph/layout.ts [hunk 7]
```diff
@@ -730,7 +828,7 @@ function createVertical(node: Node, parents?: string[], groupId?: string): Right
   };
 }
 
-function createRight(node: RightNode, parents?: string[], groupId?: string): Right {
+function createRight(node: RightNode, parents?: string[], groupId?: string, index?: number): Right {
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/layout.ts:831]:
```suggestion
function createRight(node: RightNode, parents?: string[], groupId?: string, order=0): Right {
```

File: libs/waterfall/src/lib/components/graph/layout.ts [hunk 8]
```diff
@@ -741,7 +839,7 @@ function createRight(node: RightNode, parents?: string[], groupId?: string): Rig
     percent: node.percent,
     actionName: 'prepend',
     pools: [], // TODO
-    order: 0, // TODO
+    order: index ?? 0,
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/layout.ts:842]:
```suggestion
    order,
```

File: libs/waterfall/src/lib/components/graph/layout.ts [hunk 9]
```diff
@@ -776,17 +874,17 @@ export function fromGraph(graph: readonly Node[]) {
         node.members.forEach(member => {
           if (member.type === 'vertical') {
             rights.push(createVertical(member, undefined, node.id));
-            member.members.forEach(subMember => {
-              rights.push(createRight(subMember, undefined, member.id));
+            member.members.forEach((subMember, index) => {
+              rights.push(createRight(subMember, undefined, member.id, index));
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/layout.ts:878]: for better consistency.
```suggestion
            member.members.forEach((subMember, order) => {
              rights.push(createRight(subMember, undefined, member.id, order));
```

File: libs/waterfall/src/lib/components/graph/layout.ts [hunk 10]
```diff
@@ -776,17 +874,17 @@ export function fromGraph(graph: readonly Node[]) {
...
           } else {
             rights.push(createRight(member, undefined, node.id));
           }
         });
       } else if (node.type === 'vertical') {
         rights.push(createVertical(node, parentIndex[node.id]));
-        node.members.forEach(member => {
-          rights.push(createRight(member, undefined, node.id));
+        node.members.forEach((member, index) => {
+          rights.push(createRight(member, undefined, node.id, index));
```
Comment (phpgeekfr) [inline libs/waterfall/src/lib/components/graph/layout.ts:887]: for better consistency.
```suggestion
        node.members.forEach((member, order) => {
          rights.push(createRight(member, undefined, node.id, order));
```

**Review by phpgeekfr — APPROVED** [review]

File: libs/waterfall/src/lib/components/graph/conditions/conditions.component.ts (hunk 1 above)
Comment (pldespaigne) [reply libs/waterfall/src/lib/components/graph/conditions/conditions.component.ts:19]: it's emitted directly in the html

File: libs/waterfall/src/lib/components/graph/conditions/step-item/step-item.component.html (hunk 2 above)
Comment (pldespaigne) [reply libs/waterfall/src/lib/components/graph/conditions/step-item/step-item.component.html:5]: reading the html, I find it more clear
```
<span class="spacer"></span>
```
vs this, we don't know the purpose of the span, one might be tempted to remove it thinking it's unused
```
<span></span>
```

File: libs/waterfall/src/lib/components/graph/graph.component.ts (hunk 3 above)
Comment (pldespaigne) [reply libs/waterfall/src/lib/components/graph/graph.component.ts]: there is no `createCondition()` function 🤔

**Pull request merged**